HPCC-12978 Modified stop_thor

Michael-Gardner opened the pull request:
Now deletes the $MASTER_PID_NAME file after we stop the process.
Changed the logic around killing the process to mimic the kill_process function used in other scripts, making the shutdown at most take 8 seconds (which used to be hard coded.)

Signed-off-by: Michael Gardner Michael.Gardner@lexisnexis.com

Michael-Gardner committed:
HPCC-12978 Stop_thor uses kill_process.  kill_process parameter order changed

Signed-off-by: Michael Gardner <Michael.Gardner@lexisnexis.com>

Michael-Gardner commented:
@jakesmith @xwang2713 Please review when you get time.

richardkchapman commented:
@jakesmith Please sanity-check

jakesmith commented on initfiles/componentfiles/thor/stop_thor:
```diff
@@ -47,23 +47,17 @@ if [ "$#" -lt "2" ] || [ "$2" != "keep_sentinel" ]; then
+masterpid=$( cat $MASTER_PID_NAME )
+TIMEOUT=8000
+kill $masterpid &>/dev/null
+while [ -d /proc/$masterpid -a $TIMEOUT -gt 0 ]; do
```
Same as before, but it feels like we should have a 'checkProcess(<pid>, <procesname>)' utility method in use here instead. To cater for the edge case where the $pid does exist but isn't thormaster.
i.e. check pid exists and proc name is what you expect.

(don't hold up this PR for that suggestion though)

Michael-Gardner replied:
You're right.  We have a function called kill_process in hpcc_setenv that does all this for us, except it deletes the sentinel file as well.  I'm going to alter it slightly so it can be used in this case too.

jakesmith commented on initfiles/componentfiles/thor/stop_thor:
```diff
@@ -47,23 +47,17 @@ if [ "$#" -lt "2" ] || [ "$2" != "keep_sentinel" ]; then
+  TIMEOUT=$(($TIMEOUT-200))
+done
+if [ $TIMEOUT -le 0 ]; then
+  kill -9 $masterpid &>/dev/null
```
The previous code would still loop around after a kill -9 ...
which seems a bit pointless in general.
However, it can happen I think, if the process gets into a zombie state..

So should probably detect if kill -9 fails to kill, report it and abort here.

jakesmith commented:
@Michael-Gardner - a couple of minor comments, which can be address in a separate PR.

Michael-Gardner commented:
making some quick changes per review

richardkchapman commented:
@jakesmith please re-review. Is this correct to target master?

Michael-Gardner commented:
@jakesmith ok so I've gone ahead and used the kill_process function that gets sourced out of hpcc_setenv.  I modified it so that it only removes the sentinel file if it gets 3 arguments, otherwise it just deals with the PID file passed and the timeout.  This keeps it compliant with all the other places it's already called.  I also modified it to pass a return value so where applicable we can check to ensure that the function killed the process correctly.  For stop_thor I don't think we should abort here.  I think it would be better to log the error and then try to kill off the slaves. Thoughts?

jakesmith commented on initfiles/componentfiles/thor/stop_thor:
```diff
@@ -47,22 +47,9 @@ if [ "$#" -lt "2" ] || [ "$2" != "keep_sentinel" ]; then
-    kill -9 $masterpid >& /dev/null
-    sleep 1
-  done
+kill_process ${MASTER_PID_NAME} 8
```
That's not what I was expecting.. 
previously it was just checking the pid (and in theory an unrelated process could take it whilst it was looping)
Now it only takes the process name.. I was expecting a utility method that took name and pid.. and verified the pid it was told to kill was indeed the name of the process it was told.
It may suffice, if MASTER_PID_NAME is unique.. but in general I think it would be better if checked.

jakesmith commented on initfiles/sbin/hpcc_setenv.in:
```diff
@@ -23,12 +23,15 @@
+        fi
     fi
+    PID=$1
+    TIMEOUT=$2
```
If 3 parameters then 1st is sentinel, 2nd is pid ...
if <3 , 1st is pid.. ?

I think that's confusing.

Why not optional 3rd parameter = sentinel ?

Michael-Gardner replied:
I did that so I wouldn't break the code in the other 11 init files where it's called.  But I'm going to go through and swap around the parameters, and add the change you talked about above for the process name check.

jakesmith commented:
@Michael-Gardner - a couple of comments

Michael-Gardner commented on initfiles/sbin/hpcc_setenv.in:
```diff
@@ -23,30 +23,39 @@
-    if [ $TIMEOUT -le 0 ]; then
-            kill -9 $WATCH_PID
+    COMM_CHECK=$( ps -q $WATCH_PID -o command= )
+    if [[ $TIMEOUT -le 0 ]] && [[ $COMM == $COMM_CHECK ]]; then
```
So now what we do is pull out the full command used to run the process from ps, then attempt to kill.  If we fail to kill before timeout hits zero, we then grab the command for the PID again and check to ensure it's the same as before.  If it is we do a kill -9 on the process.  We return 1 if kill -9 doesn't work, and 0 if everything works fine.

xwang2713 replied:
Should be "ps -p <pid> -o command="? Also the output has two column, for example "/var/lib/HPCCSystems/mythor/thormaster_mythor MASTER=10.176.152.75:2000". Does it need retrieve the first one?

Michael-Gardner replied:
yes it should be -p. All that matters is that the entire string from COMM is the same as what we get from COMM_CHECK, so multiple columns is fine.

Michael-Gardner commented:
@xwang2713 Please take another look at the changes when you get a chance.  Touched a wee bit of code since you last saw it.

Michael-Gardner committed:
HPCC-12978 Stop_thor uses kill_process.  kill_process parameter order changed

Signed-off-by: Michael Gardner <Michael.Gardner@lexisnexis.com>

Michael-Gardner commented:
@jakesmith ready for you to look at.  Should have the changes we talked about on flowdock.

jakesmith commented on initfiles/bin/init_thor:
```diff
@@ -29,7 +29,7 @@ rm -f ${SENTINEL}
         echo "Stopping"
         $deploydir/stop_thor $deploydir
-        kill_process ${SENTINEL} ${PID_NAME} 3
+        kill_process ${PID_NAME} thormaster 3 ${SENTINEL}
```
I don't think 'thormaster' is enough..
The process name run is a softlink to thormaster_lcr, that looks like e.g. thormaster_mythor

It's normal to have >1 thormaster on the same physical node in some environments.
So grepping for 'thormaster' would not uniquely identify the correct one.

Michael-Gardner commented:
@jakesmith tested passing thormaster_${comp_base} and I also removed the kill_process call from init_thor because the killed() function calls stop_thor which in turn calls kill_process itself.

jakesmith commented on initfiles/sbin/hpcc_setenv.in:
```diff
@@ -23,30 +23,46 @@
-            kill -9 $WATCH_PID
+
+    if [[ $TIMEOUT -le 0 ]] && ps -p $WATCH_PID -o command= | grep "${PROCESS}" &>/dev/null ; then
+      kill -9 $WATCH_PID &>/dev/null
```
may be worth some logging here.. e.g. failed to kill gracefully..

jakesmith commented on initfiles/sbin/hpcc_setenv.in:
```diff
@@ -23,30 +23,46 @@
+    local TIMEOUT=$(($3*1000))
+
+    if ps -p $WATCH_PID -o command= | grep -v "${PROCESS}" &>/dev/null ; then
+      exit 1
```
hm, that means if WATCH_PID exists, but it's not ${PROCESS), then abort the script with an error?

If another process happened to have started to use PID, then that's okay.. it just means 'our' process has quit and OS has reused the PID # .. should just return success here I think.

jakesmith commented on initfiles/sbin/hpcc_setenv.in:
```diff
@@ -23,30 +23,46 @@
+      kill -0 $WATCH_PID &>/dev/null
+      # if still alive
+      if [[ $? -eq 0 ]]; then
+        return 1
```
Does anything check return result of pidwait_fn() ?

Didn't it used to do something different here, i.e. have detected zombie.. 
Am not really sure what best thing to do is here.. if process won't die.., probably just log it.
It might be more pertinent to detect and abort out on start if there's an existing process around..

Michael-Gardner committed:
HPCC-12978 Made changes in line with PR

Signed-off-by: Michael Gardner <Michael.Gardner@lexisnexis.com>

jakesmith commented:
Can't see any other issues.
Good to merge.
@richardkchapman
